distributed/comm/ucx.py

```diff
@@ -325,10 +325,10 @@ async def read(self, deserializers=("cuda", "dask", "pickle", "error")):
             await self.ep.recv(header)
             header = struct.unpack(header_fmt, header)
             cuda_frames, sizes = header[:nframes], header[nframes:]
-        except (
-            ucp.exceptions.UCXCloseError,
-            ucp.exceptions.UCXCanceled,
-        ) + (getattr(ucp.exceptions, "UCXConnectionReset", ()),):
+        except BaseException:
+            # In addition to UCX exceptions, may be CancelledError or a another
+            # "low-level" exception. The only safe thing to do is to abort.
+            # (See also https://github.com/dask/distributed/pull/6574).
             self.abort()
             raise CommClosedError("Connection closed by writer")
         else:
```

Comment:
I think we also need to catch connection issues on line 354 as well.

So perhaps lines 353 and 354 should be replaced by:
```python
try:
    for frame in recv_frames:
        await self.ep.recv(frame)
except BaseException as e:
    raise CommClosedError("Connection closed by writer.\nInner exception: {e!r}")
```

I had thought that one might be able to reduce synchronisation a _little_ bit by using:

```
await asyncio.gather(*(map(self.ep.recv, recv_frames))
```

With a matching change in `write` of `await asyncio.gather(*map(self.ep.send, send_frames))`.

But I am unsure of the semantics of UCX wrt message overtaking. I think this could potentially result in the second (say) sent frame ending up in the first receive slot, which would be bad.

Reply:
> I think we also need to catch connection issues on line 354 as well.
> 
> So perhaps lines 353 and 354 should be replaced by:
> 
> ```python
> try:
>     for frame in recv_frames:
>         await self.ep.recv(frame)
> except BaseException as e:
>     raise CommClosedError("Connection closed by writer.\nInner exception: {e!r}")
> ```

I'm not entirely sure we want that, maybe it never occurred in practice or just raising the original exception may be fine. I'm mostly concerned with unforeseen side-effects this may cause and would prefer not to mess with it now given it's not been a problem so far. WDYT?

> I had thought that one might be able to reduce synchronisation a _little_ bit by using:
> 
> ```
> await asyncio.gather(*(map(self.ep.recv, recv_frames))
> ```
> 
> With a matching change in `write` of `await asyncio.gather(*map(self.ep.send, send_frames))`.
> 
> But I am unsure of the semantics of UCX wrt message overtaking. I think this could potentially result in the second (say) sent frame ending up in the first receive slot, which would be bad.

I would expect that as well and had done it once and had to revert https://github.com/dask/distributed/pull/5505 because that caused various issues, unfortunately. In any case, with the C++ UCX introduction of "multi-transfers" this will anyway be reduced to a single future, so I will not try to improve this code in its current form.

Reply:
> I'm not entirely sure we want that, maybe it never occurred in practice or just raising the original exception may be fine. I'm mostly concerned with unforeseen side-effects this may cause and would prefer not to mess with it now given it's not been a problem so far. WDYT?

What could happen (although my guess is that it would be low likelihood) is that we're receiving a bunch of frames, each `await` yields to the event loop, and in between `awaits` the remote endpoint is closed for some other reason.

Reply:
Yes, but I fear that by raising a different exception now we may end up in some different control path that we didn't expect. I'm hoping that this patch can end up in the next Distributed release and it could be included in RAPIDS 22.10. I would be fine trying that out afterwards, but I'm a bit nervous of breaking something close to release time.

Reply:
OK, thanks, makes sense.